V3: type hints for operations

stone-w4tch3r opened the pull request:
Type hints for all operations, V3 version

stone-w4tch3r committed:
more operations typing

stone-w4tch3r committed:
more typing in operations

stone-w4tch3r committed:
typing in server.py

stone-w4tch3r committed:
operations typing done!

stone-w4tch3r committed:
pre-pr review

stone-w4tch3r committed:
failed tests fixed

stone-w4tch3r committed:
postgres typing

stone-w4tch3r committed:
linting fixed

stone-w4tch3r commented:
Looks like some check are broken :(

stone-w4tch3r committed:
type hints update for files.put

Fizzadar commented:
@stone-w4tch3r thank you so much for working on this! At a quick glance it looks fantastic, time is in extremely short supply right now I won’t be able to properly review till next weekend unfortunately, just a heads up!

Fizzadar commented on pyinfra/operations/selinux.py:
```diff
@@ -25,16 +39,19 @@ def boolean(bool_name, value, persistent=False):
             name='Allow Apache to connect to LDAP server',
             'httpd_can_network_connect',
-            'on',
+            Boolean.ON,
```
```suggestion
            selinux.Boolean.ON,
```

Fizzadar commented on pyinfra/operations/selinux.py:
```diff
@@ -127,12 +144,15 @@ def port(protocol, port_num, se_type=None, present=True):
         selinux.port(
             name='Allow Apache to provide service on port 2222',
-            'tcp',
+            Protocol.TCP,
```
```suggestion
            selinux.Protocol.TCP,
```

Fizzadar reviewed (commented):
OK that week turned into a month, my apologies @stone-w4tch3r. PR looks great, HUGE thank you for working on this!

I think the failures are mostly missing `Optional`s since mypy switched to disallowing implicit optional kwargs. Looks like there's [a tool](https://github.com/hauntsaninja/no_implicit_optional) to automatically fix this.

I can clean those up and get that working with the tests, all the types look good 👍

Fizzadar committed:
typing started

# Conflicts:
#	pyinfra/facts/util/packaging.py
#	pyinfra/operations/apt.py

Fizzadar committed:
improved typing in rpm-based operations

# Conflicts:
#	pyinfra/operations/dnf.py
#	pyinfra/operations/util/packaging.py
#	pyinfra/operations/yum.py

Fizzadar committed:
typing in files

# Conflicts:
#	pyinfra/operations/files.py
#	pyinfra/operations/util/files.py

Fizzadar committed:
gem & git typing

# Conflicts:
#	pyinfra/operations/gem.py
#	pyinfra/operations/git.py

Fizzadar committed:
Fix arg name, formatting

Fizzadar commented:
Rebased ontop of v3, hoping to get remaining mypy errors cleaned up this week.

Fizzadar committed:
Final pass over all the broken type issues

Mostly just adding `| None` to all the optional operation arguments.

Fizzadar committed:
Add dev lint/test scripts

Fizzadar committed:
Future annotations imports

Fizzadar reviewed (approved):
Got there in the end! Massive thank you @stone-w4tch3r for doing the majority of the work here, truly a fantastic achievement to land this in v3!
